src/coreclr/inc/corinfo.h

```diff
@@ -2354,6 +2354,9 @@ class ICorStaticInfo
     // Quick check whether the type is a value class. Returns the same value as getClassAttribs(cls) & CORINFO_FLG_VALUECLASS, except faster.
     virtual bool isValueClass(CORINFO_CLASS_HANDLE cls) = 0;
 
+    // Checks if type can be compared for equality as bytes.
+    virtual bool isBitwiseEquatable(CORINFO_CLASS_HANDLE cls) = 0;
```

Comment:
I have mixed feeling about introducing a new JIT/EE interface just to support an expansion of one-off BCL method in the JIT. Providing the expansion as IL on the VM is less spread through the system.

This direction makes sense only if we believe that 100% of the IL intrinsics are going to move to the JIT eventually. We would need to have a number of new one-off methods like this to pull it off.

@dotnet/jit-contrib Opinions?

Reply:
I'd say that the this proper recognition of this intrinsic as a constant in the JIT is important since it makes sure it can properly skip importing branches guarded by it, similar to how `IsKnownConstant` and `IsReferenceOrContainsReferences` are used. It's also tracked for exposal in #75642.

As far as I could see, the only IL intrinsics left in the CoreCLR VM are the Unsafe ones (tracked for removal in #69220), `Interlocked.CompareExchange(T)` which can be removed with no impact today since it already has an equivalent managed implementation with `Unsafe.As` and `Activator.CreateInstance<T>` which seems to be the only problematic one.

Since "bitwise equality" is a somewhat fundamental type property (and could potentially be useful for other optimizations in the JIT), I've considered making this a flag in `getClassAttribs`, but I've refrained from doing so since it's I think relatively expensive to fetch and wouldn't be used that often.
However, citing @SingleAccretion from Discord:
```
As Tanner notes, it may be good to design a more general mechanism to ask questions about facts like that.
For example, an API that takes flags with the "requests" and returns answers to only the questions asked.
```
maybe it'd make sense to introduce `getExtendedClassAttribs(uint32_t mask)` that'd only fetch more rarely used/expensive flags (with a new separate enum for them) like this or #88136 based on the mask?

Reply:
> This direction makes sense only if we believe that 100% of the IL intrinsics are going to move to the JIT eventually

Do we have a list of the APIs that would still require moving and how complex they are to handle? That is, are many of them effectively simple flags like this one which are often trivial to lookup and which may get used on generic hot paths or do we have a number of substantially more complex APIs where the benefits of special-casing them in the JIT are more questionable?

Reply:
> it can properly skip importing branches guarded by it, similar to how IsKnownConstant and IsReferenceOrContainsReferences are used

FWIW, this is a workaround for a deficiency in the current JIT inlining strategy. It would be better to fix it in more general way for everybody out there instead of using it to justify implementing more methods as JIT intrinsics.

> Interlocked.CompareExchange(T) which can be removed with no impact today

I expect that we would see performance regressions in shared generic code callsites. It is a workaround for the current shared generic code inlining limitation.

> maybe it'd make sense to introduce getExtendedClassAttribs(uint32_t mask)

`getExtendedClassAttribs(uint32_t mask)` would make sense if the typically caller needs most of the flags and when there is an efficiency benefit from computing and returning multiple flags together. If the typically caller needs just one flag as is the case for expanding one-off Intrinsics, it is better to have an API that returns the one property directly and avoids the extra level of abstraction.

Reply:
> Do we have a list of the APIs that would still require moving and how complex they are to handle?

The current CoreCLR and native AOT intrinsics that would required introducing new special new JIT-EE interface APIs. You can generally assume that it is one new JIT/EE interface method for each:

- `Activaror.CreateInstanceOf<T>`
- `EqualityCompaparer/Comparer.Create<T>`
- `EqualityComparerHelpers.GetComparerForReferenceTypesOnly/StructOnlyEquals`
- `Interlocked.CompareExchange<T>(ref)`
- `RuntimeAugments.GetCanonType`
- `MethodTable.SupportsRelativePointers`
- `System.IO.Stream.HasOverriddenBeginEndRead/HasOverriddenBeginEndReadHasOverriddenBeginEndWrite`

We like introducing intrinsics to do various tricks, so this list would grow over time. Also, some of these would require special-handling in the native AOT scanner if they are not IL intrinsics anymore.

Reply:
> FWIW, this is a workaround for a deficiency in the current JIT inlining strategy. It would be better to fix it in more general way for everybody out there instead of using it to justify implementing more methods as JIT intrinsics.

Even then, having just a temporary workaround is better than not having one.

> I expect that we would see performance regressions in shared generic code callsites. It is a workaround for the current shared generic code inlining limitation.

I don't really think that's the case, I think the only difference is the smaller IL size due to no Unsafe.As calls and the inlining boost from being an intrinsic.
https://github.com/dotnet/runtime/blob/31106939ba524a7e6df8225cf8d0fedefba92b9f/src/coreclr/System.Private.CoreLib/src/System/Threading/Interlocked.CoreCLR.cs#L120-L128

> The current CoreCLR and native AOT intrinsics that would required introducing new special new JIT-EE interface APIs. You can generally assume that it is one new JIT/EE interface method for each

I think that it only makes sense to move apis that are: performance sensitive (like the three `Is` APIs here), more valid (like the `GetMethodTable` one) or ones that have big implementations in both runtimes (like the enum ones). I don't think it makes much sense to have runtime specific, rarely used intrinsics like `HasOverriddenBeginEndRead` that'd just call into the VM.

Reply:
> having just a temporary workaround is better than not having one.

Can you come up with a real-world looking example where `isBitwiseEquatable` makes a difference to show that it is worth doing?

> more valid (like the GetMethodTable one) 

Yes, the GetMethodTable one looks good.

> ones that have big implementations in both runtimes (like the enum ones)

The implementations in each runtime are about 30 lines of code each. I would not call that big. The inlining expansion in the JIT is not reducing runtime complexity much overall.

The expansion in the JIT makes the code less maintainable by creating a copy of CompareTo implementation from 10 different .cs files into the JIT. To make this explicit, you would want to add a comment "This logic is duplicated in the JIT in EnumCompare intrinsic expansion" to CompareTo implementation in all Byte.cs, UByte.cs, ...., etc. I do not think it would be a good idea.

Reply:
> > >  Interlocked.CompareExchange(T) which can be removed with no impact today

> > I don't really think that's the case, I think the only difference is the smaller IL size due to no Unsafe.As calls and the inlining boost from being an intrinsic.

> I don't really think that's the case, I think the only difference is the smaller IL size due to no Unsafe.As calls and the inlining boost from being an intrinsic.

I have done a few quick tests and I agree that the IL intrinsic for Interlocked.CompareExchange does not seem to be making a difference. Feel free to remove it.

Reply:
> @dotnet/jit-contrib Opinions?

I don't understand the motivation for these changes. It seems like a lot of code churn late in a release cycle with a decent risk of introducing correctness or perf issues.

If we think further unification / simplification here is sufficient motivation, perhaps it can wait until .NET 9?